app/code/Magento/Catalog/Model/CategoryRepository.php

```diff
@@ -265,4 +265,19 @@ public function _resetState(): void
     {
         $this->instances = [];
     }
+
+    /**
+     * Get store id for category
+     *
+     * @param CategoryInterface $category
+     * @return int
+     */
+    private function getCategoryStoreId($category)
+    {
+        $storeId = (int)$this->storeManager->getStore()->getId();
+        if ($storeId > 0 && $category->hasData('store_id')) {
+            $storeId = (int)$category->getStoreId();
+        }
+        return $storeId;
+    }
```

Comment:
Maybe this one will be better? We need to firstly check then category has store id and if yes then we return it.

In other case as a fallback we return default one from store manager. 

```suggestion
    private function getCategoryStoreId($category)
    {
        if ($category->hasData('store_id')) {
            return (int)$category->getStoreId();
        }
        
        return (int)$this->storeManager->getStore()->getId();
    }
```

Reply:
it same first commit https://github.com/magento/magento2/pull/38222/commits/6b24a6e5e27aef5fc7e6191a5465b89d134c4324#diff-de856f314742e09493164497722b3a2988169e277c96320be22c0bce739cc073 but it make fail many case integration test